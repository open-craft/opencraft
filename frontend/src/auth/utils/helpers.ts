import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
