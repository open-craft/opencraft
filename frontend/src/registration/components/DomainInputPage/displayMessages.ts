const messages = {
    typeDomainNameBelow: {
        defaultMessage: 'Type your domain name below',
        description: 'Message above domain input text box.',
    },
    checkAvailability: {
        defaultMessage: 'Check Availability',
        description: 'Label for button that check availability of Domain',
    },
    useOwnDomain: {
        defaultMessage: 'I want to use my own domain',
        description: 'Link to direct user to page where they can set up their own domain.',
    },
};

export default messages;
