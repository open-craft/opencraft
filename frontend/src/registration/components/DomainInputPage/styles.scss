@import '~styles/theme';

.div-fill {
  height: 100%;
}

.use-own {
  text-align: center;
  cursor: pointer; 
  text-decoration: underline;
  font-family: $chivo-font;
  font-size: 18px;
  letter-spacing: 0.54px;
  color: #1e4c59;
  margin-top: 40px;
}
