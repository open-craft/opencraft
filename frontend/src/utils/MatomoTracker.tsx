import React, { ReactNode } from 'react';
import {
  createInstance,
  MatomoProvider,
  useMatomo
} from '@datapunt/matomo-tracker-react';
import {
  MATOMO_ALIAS_DOMAIN,
  MATOMO_BASE_URL,
  MATOMO_MY_DOMAIN,
  MATOMO_SITE_ID
} from '../global/constants';

interface MatomoTrackerProps {
  children: ReactNode;
}

interface MatomoAdditionalConfiguration {
  [propName: string]: any;
}

interface MatomoConfiguration {
  urlBase: string;
  siteId: number;
  configurations?: MatomoAdditionalConfiguration;
}

interface MatomoUserIdTrackerProps {
  userId: string;
}

/*
 * This wrapper component initializes the Matomo tracking by wrapping the children
 * elements, if the tracking is enabled by providing a valid value for the 'MATOMO_BASE_URL'
 * constant. It allows tracking page views, events and the user id, when available.
 *
 * The 'MATOMO_ALIAS_DOMAIN' constant allows the cross-domain linking of another site with
 * this site so that they are treated as the same site in Matomo.
 */
export const MatomoTracker: React.FC<MatomoTrackerProps> = (
  props: MatomoTrackerProps
) => {
  let element = props.children;
  if (MATOMO_BASE_URL) {
    const matomoConfiguration: MatomoConfiguration = {
      urlBase: MATOMO_BASE_URL,
      siteId: parseInt(MATOMO_SITE_ID, 10)
    };
    if (MATOMO_ALIAS_DOMAIN) {
      const myDomain = `*.${MATOMO_MY_DOMAIN}`;
      const aliasDomain = `*.${MATOMO_ALIAS_DOMAIN}`;
      matomoConfiguration.configurations = {
        enableCrossDomainLinking: true,
        setDomains: [myDomain, aliasDomain]
      };
    }
    element = (
      <MatomoProvider value={createInstance(matomoConfiguration)}>
        {props.children}
      </MatomoProvider>
    );
  }
  return <>{element}</>;
};

/**
 * This is a component passes the provided user id to Matomo if Matomo tracking is enabled.
 * The 'pushInstruction' function exposed by the 'useMatomo()' hook is used to do this.
 *
 * This component is needed as all the components in the registration process are classes and hence
 * the 'yseMatomo' hook cannot be directly used in those components.
 */
export const MatomoUserIdTracker: React.FC<MatomoUserIdTrackerProps> = (
  props: MatomoUserIdTrackerProps
) => {
  const { pushInstruction } = useMatomo();
  if (MATOMO_BASE_URL && props.userId) {
    pushInstruction('setUserId', props.userId);
  }
  return null;
};
