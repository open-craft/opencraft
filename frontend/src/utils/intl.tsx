import React from "react";
import { FormattedMessage, MessageDescriptor } from "react-intl";

type MessageMap = { [k: string]: Omit<MessageDescriptor, 'id'> };

interface Props<T, K extends keyof T> {
    id: K;
    messages: T;
    values?: any;
}


export const WrappedMessage = (props: Props<MessageMap, string>) => {
    const {messages, id, ...values} = props;
    return (
        <FormattedMessage  {...messages[id]} id={id} {...values}/>
    );
};
