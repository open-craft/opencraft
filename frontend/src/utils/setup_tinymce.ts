import 'tinymce/tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';
import 'tinymce/skins/ui/oxide/content.min.css';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/print';
import 'tinymce/plugins/preview';
import 'tinymce/plugins/anchor';
import 'tinymce/plugins/searchreplace';
import 'tinymce/plugins/visualblocks';
import 'tinymce/plugins/code';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/table';
import 'tinymce/plugins/paste';
import 'tinymce/plugins/help';
import 'tinymce/plugins/wordcount';
