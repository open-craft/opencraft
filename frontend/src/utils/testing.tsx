import { createMemoryHistory } from 'history';
import * as React from "react";
import { IntlProvider } from "react-intl";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import * as renderer from 'react-test-renderer';
import { createStore } from "redux";
import { createRootReducer } from "../global/reducers";

export const setupComponentForTesting = (reactContent: JSX.Element, storeContents = {}) => {
    const store = createStore(createRootReducer(createMemoryHistory()), storeContents);
    return renderer.create(
        <IntlProvider textComponent={React.Fragment} locale="en">
            <Provider store={store}>
                <MemoryRouter>
                    {reactContent}
                </MemoryRouter>
            </Provider>
        </IntlProvider>
    );
};
