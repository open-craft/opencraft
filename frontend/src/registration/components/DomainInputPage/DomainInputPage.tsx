import { RegistrationSteps } from 'global/constants';
import { RootState } from 'global/state';
import * as React from 'react';
import { connect } from 'react-redux';
import { WrappedMessage } from 'utils/intl';
import { DomainInput, InstitutionalAccountHero } from 'ui/components';
import { RedirectToCorrectStep } from 'registration/components';
import { RegistrationStateModel } from 'registration/models';
import { Nav } from 'react-bootstrap';
import { performValidationAndStore, clearErrorMessage } from '../../actions';
import { RegistrationPage } from '../RegistrationPage';
import messages from './displayMessages';
import './styles.scss';

interface ActionProps {
  performValidationAndStore: Function;
  clearErrorMessage: Function;
}

interface State {
  subdomain: string;
  externalDomain: string;
  domainIsExternal: boolean;
}

interface Props extends ActionProps {}
interface StateProps extends RegistrationStateModel {}

interface Props extends StateProps, ActionProps {}

@connect<{}, ActionProps, {}, Props, RootState>(
  (state: RootState) => ({
    ...state.registration
  }),
  {
    performValidationAndStore,
    clearErrorMessage
  }
)
export class DomainInputPage extends React.PureComponent<Props, State> {
  constructor(props: Props, state: State) {
    super(props);

    this.state = {
      subdomain: props.registrationData.subdomain,
      externalDomain: props.registrationData.externalDomain,
      domainIsExternal: false
    };
  }

  private handleDomainChange = (newDomain: string) => {
    this.setState({
      subdomain: newDomain
    });
    // Clean up error feedback if any
    if (this.props.registrationFeedback.subdomain) {
      this.props.clearErrorMessage('subdomain');
    }
  };

  private handleExternalDomainChange = (newDomain: string) => {
    this.setState({
      externalDomain: newDomain
    });
    // Clean up error feedback if any
    if (this.props.registrationFeedback.subdomain) {
      this.props.clearErrorMessage('externalDomain');
    }
  };

  private submitDomain = () => {
    if (this.state.domainIsExternal) {
      let subdomain = this.state.externalDomain.replace('.', '');
      // Append random digits to internal domain to avoid conflicts
      subdomain = `${subdomain}-${Math.random()
        .toString(36)
        .substr(2, 4)}`;
      this.props.performValidationAndStore(
        {
          subdomain,
          externalDomain: this.state.externalDomain
        },
        RegistrationSteps.CUSTOM_DOMAIN
      );
    } else {
      this.props.performValidationAndStore(
        {
          subdomain: this.state.subdomain
        },
        RegistrationSteps.INSTANCE
      );
    }
  };

  private handleSwitchPageToExternal = (domainIsExternal: boolean) => {
    this.setState({ domainIsExternal });
  };

  private renderInternalDomain = () => {
    return (
      <div className="div-fill">
        <RegistrationPage
          title="Pro & Teacher Account"
          subtitle="Create your own Open edX instance now."
          currentStep={1}
        >
          <RedirectToCorrectStep
            currentPageStep={0}
            currentRegistrationStep={this.props.currentRegistrationStep}
          />
          <DomainInput
            domainName={this.state.subdomain}
            error={this.props.registrationFeedback.subdomain}
            internalDomain
            loading={this.props.loading}
            handleDomainChange={this.handleDomainChange}
            handleSubmitDomain={this.submitDomain}
          />
          <div className="use-own">
            <Nav.Link
              onClick={() => {
                this.handleSwitchPageToExternal(true);
              }}
            >
              <WrappedMessage messages={messages} id="useOwnDomain" />
            </Nav.Link>
          </div>
        </RegistrationPage>
        <InstitutionalAccountHero />
      </div>
    );
  };

  private renderExternalDomain = () => {
    return (
      <div className="div-fill">
        <RegistrationPage
          title="Pro & Teacher Account"
          subtitleBig="Register with your own domain"
          subtitle="Cost: +â‚¬25/month"
          currentStep={1}
        >
          <RedirectToCorrectStep
            currentPageStep={0}
            currentRegistrationStep={this.props.currentRegistrationStep}
          />
          <DomainInput
            domainName={this.state.externalDomain}
            error={this.props.registrationFeedback.externalDomain}
            internalDomain={false}
            loading={this.props.loading}
            handleDomainChange={this.handleExternalDomainChange}
            handleSubmitDomain={this.submitDomain}
          />
          <div className="use-own">
            <Nav.Link
              onClick={() => {
                this.handleSwitchPageToExternal(false);
              }}
            >
              <WrappedMessage messages={messages} id="useInternalDomain" />
            </Nav.Link>
          </div>
        </RegistrationPage>
        <InstitutionalAccountHero />
      </div>
    );
  };

  public render() {
    if (this.state.domainIsExternal) {
      return this.renderExternalDomain();
    }

    return this.renderInternalDomain();
  }
}
