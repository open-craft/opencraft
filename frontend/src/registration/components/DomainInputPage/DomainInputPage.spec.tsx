import React from 'react';
import { setupComponentForTesting } from 'utils/testing';
import { DomainInputPage } from './DomainInputPage';

describe("Domain Input Page", function() {
  it('Correctly renders domain input page', () => {
      const tree = setupComponentForTesting(
        <DomainInputPage />,
        {
          registration: {
            registrationData: {
              subdomain: 'test-subdomain',
              externalDomain: ''
            },
            registrationFeedback: {}
          }
        }
      ).toJSON();
      expect(tree).toMatchSnapshot();
  });

  it('Correctly renders domain input with error fedback', () => {
      const tree = setupComponentForTesting(
        <DomainInputPage />,
        {
          registration: {
            registrationData: {
              subdomain: 'test-subdomain',
              externalDomain: ''
            },
            registrationFeedback: {
              subdomain: ['subdomain test error']
            }
          }
        }
      ).toJSON();
      expect(tree).toMatchSnapshot();
  });

  it('Correctly goes to custom domain page and back', () => {
      let component = setupComponentForTesting(
        <DomainInputPage />,
        {
          registration: {
            registrationData: {
              subdomain: 'test-subdomain',
              externalDomain: ''
            },
            registrationFeedback: {}
          }
        }
      )

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      // Click on custom domain page
      component.root.findByType("a").props.onClick({preventDefault: () => {}});
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      // Click on opencraft domain page
      component.root.findByType("a").props.onClick({preventDefault: () => {}});
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });
});
