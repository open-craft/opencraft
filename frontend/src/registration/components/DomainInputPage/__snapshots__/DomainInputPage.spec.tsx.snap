// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<form
  className=""
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="domainNameInput"
    >
      Type your domain name below
    </label>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        defaultValue=""
        id="domainNameInput"
        onChange={[Function]}
        placeholder="yourdomain"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Check Availability
        </button>
      </div>
    </div>
  </div>
  <a
    href="#"
  >
    I want to use my own domain
  </a>
</form>
`;
